libr/include/r_arch.h

```diff
@@ -0,0 +1,100 @@
+/* radare2 - LGPL - Copyright 2020 - pancake */
+
+#ifndef R2_ARCH_H
+#define R2_ARCH_H
+
+#include <r_util.h>
+#include <r_bind.h>
+#include <r_io.h> // Just for RIOBind which is needed for some plugins
+
+typedef int RArchBits;
+typedef int RArchEndian;
+
+typedef enum r_arch_decode_options_t {
+	R_ARCH_OPTION_CODE = 1 << 0, // 1,
+	R_ARCH_OPTION_SIZE = 1 << 1, // 2,
+	R_ARCH_OPTION_ANAL = 1 << 2, //4,
+	R_ARCH_OPTION_ESIL = 1 << 3, //8,
+	R_ARCH_OPTION_OPEX = 1 << 4, //16,
+	R_ARCH_OPTION_DESC = 1 << 5, // 32, // instruction description
+} RArchOptions;
+
+typedef struct r_arch_t {
+	struct r_arch_plugin_t *cur;
+	RArchEndian endian;
+	RArchBits bits;
+	char *cpu;
+	RList *plugins;
+	RIOBind iob;
+	size_t syntax;
+} RArch;
+
+typedef struct r_arch_info_t {
+	size_t align;
+} RArchInfo;
+
+typedef struct r_arch_instruction_t {
+	ut64 addr;
+	size_t size; // inherit from thisi.data.len?
+	ut32 opid; // opcode identifier
+	ut64 type; // opcode type
+	RStrBuf code; // disasm
+	RStrBuf data; // bytes
+	RStrBuf esil; // emulation
+	RStrBuf opex; // analysis
+	size_t syntax; // used for disasm
+	RVector dest; // array of destinations
+} RArchInstruction;
+
+
+typedef bool (*RArchEncodeCallback)(RArch *a, RArchInstruction *ins, RArchOptions options);
+typedef bool (*RArchDecodeCallback)(RArch *a, RArchInstruction *ins, RArchOptions options);
+typedef char * (*RArchRegistersCallback)(RArch *a);
+typedef void (*RArchInfoCallback)(RArch *a);
+
+typedef struct r_arch_plugin_t {
+	const char *name;
+	const char *arch;
+	const char *cpus;
+	const char *features;
+	RArchBits bits;
+	RArchEndian endian;
+	// copyright
+	const char *author;
+	const char *desc;
+	const char *license;
+	const char *version;
+	// callbacks
+	RArchEncodeCallback encode;
+	RArchDecodeCallback decode;
+	RArchRegistersCallback registers;
+	RArchInfoCallback info; // dynamic info
+	bool (*init)(void *user);
+	bool (*fini)(void *user);
+} RArchPlugin;
+
+R_API RArchInstruction *r_arch_instruction_new();
+R_API void r_arch_instruction_free(RArchInstruction *ins);
+R_API void r_arch_instruction_init(RArchInstruction *ins);
+R_API void r_arch_instruction_init_data(RArchInstruction *ins, ut64 addr, const ut8 *buf, size_t len);
+R_API void r_arch_instruction_init_code(RArchInstruction *ins, ut64 addr, const char *opstr);
+R_API void r_arch_instruction_fini(RArchInstruction *ins);
+
+R_API RArch *r_arch_new();
+R_API void r_arch_free(RArch *arch);
+
+R_API bool r_arch_encode(RArch *a, RArchInstruction *ins, RArchOptions opt);
```

Comment:
`r_arch_encode` always uses the `ins.code`, right? I don't really like the fact that you have somehow to know that you have to setup ins.code before calling this function. I think I would prefer something like:
```C
ins = r_arch_decode (arch, "nop", opt);
```

Reply:
actually, this is not decoding, it would  be encoding.

this is how the intiial design was, and there was a bunch of limitations, thats why i decided to go for passing an instruction instead of just the instruction to **encode**

Reply:
what were the limitations? IMHO it is confusing having to set just few parts of a struct when you actually just care about the code field. Just directly pass that each time, it makes the API much clearer.